spec/components/discourse_redis_spec.rb

```diff
@@ -0,0 +1,64 @@
+require 'rails_helper'
+
+describe DiscourseRedis do
+  let(:slave_host) { 'testhost' }
+  let(:slave_port) { 1234 }
+
+  let(:config) do
+    DiscourseRedis.config.dup.merge({
+      slave_host: 'testhost', slave_port: 1234, connector: DiscourseRedis::Connector
+    })
+  end
+
+  let(:fallback_handler) { DiscourseRedis::FallbackHandler.instance }
+
+  context '.slave_host' do
+    it 'should return the right config' do
+      slave_config = DiscourseRedis.slave_config(config)
+      expect(slave_config[:host]).to eq(slave_host)
+      expect(slave_config[:port]).to eq(slave_port)
+    end
+  end
+
+  describe DiscourseRedis::Connector do
+    let(:connector) { DiscourseRedis::Connector.new(config) }
+
+    it 'should return the master config when master is up' do
+      expect(connector.resolve).to eq(config)
+    end
+
+    it 'should return the slave config when master is down' do
+      begin
+        Redis::Client.any_instance.expects(:call).raises(Redis::CannotConnectError).twice
+        expect { connector.resolve }.to raise_error(Redis::CannotConnectError)
+
+        config = connector.resolve
+
+        expect(config[:host]).to eq(slave_host)
+        expect(config[:port]).to eq(slave_port)
+      ensure
+        fallback_handler.master = true
+      end
+    end
+  end
+
+  describe DiscourseRedis::FallbackHandler do
+    it 'should switch back to the master server' do
+      begin
+        fallback_handler.master = false
+        $redis.without_namespace.expects(:get).raises(Redis::CommandError.new("READONLY"))
+        Redis::Client.any_instance.expects(:call).with([:info]).returns(DiscourseRedis::FallbackHandler::MASTER_LINK_STATUS)
+        Redis::Client.any_instance.expects(:call).with([:client, [:kill, 'type', 'normal']])
+
+        current_threads = Thread.list
+        expect { $redis.get('test') }.to change { Thread.list.count }.by(1)
```

Comment:
this is not really a good test ... its testing for a side effect of a very particular implementation. this really should not matter.